### ISSUE TITLE
Question about default values in hook implementation

### ISSUE THREAD

**Issue by krassowski** [pytest-dev/pluggy#442]
I encountered a somewhat confusing behaviour in https://github.com/python-lsp/python-lsp-server/issues/439 were I observe that the hook implementation default arguments always take precedence over the values passed to the hook at runtime. This can be reproduced using the following snippet:

```python
import pluggy

hookspec = pluggy.HookspecMarker("test")
hookimpl = pluggy.HookimplMarker("test")

class MySpec:

    @hookspec
    def myhook(self, arg1, arg2):
        pass

class Plugin_1:

    @hookimpl
    def myhook(self, arg1, arg2=0):
        return arg1 + arg2

pm = pluggy.PluginManager("test")
pm.add_hookspecs(MySpec)
pm.register(Plugin_1())
results = pm.hook.myhook(arg1=1, arg2=2)
print(results)
```

which prints `[1]` rather than `[3]` which I would expect. I see this issue in Python 3.8 and 3.11.

I do not see any explanation of this behaviour, which appears highly unexpected on https://pluggy.readthedocs.io/en/latest/ but I also do not see any hook implementations with default values there either. As such I am confused as to whether this is intended behaviour or not. Which is it?

**Comment by RonnyPfannschmidt** [comment]
As of now There's no support for default values at all, thus never use them in Hook impls

**Comment by bluetech** [comment]
I agree this is not unexpected behavior. I think values passed by the caller should take precedence over default specified by the hookimpl.

I am working separately on allowing *hookspecs* to declare defaults, which also interacts with this, but that's a separate matter.

**Comment by bluetech** [comment]
Hookspec defaults are merged now (#732) - not yet released at time of writing.

So now an argument value can come from 3 places: spec default, impl default, call:

```py
import pluggy

pm = pluggy.PluginManager("foo")
hookspec = pluggy.HookspecMarker("foo")
hookimpl = pluggy.HookimplMarker("foo")

class Spec:
    @hookspec
    def meth(self, arg="spec"):
        pass

pm.add_hookspecs(Spec)

class Plugin:
    @hookimpl
    def meth(self, arg="impl"):
        return arg

pm.register(Plugin())
print(pm.hook.meth(arg="call"))
```

The current precedence order is (higher = preferred):

1. impl default
2. call
3. spec default

I think it should be:

1. call
2. spec default
3. impl default

The first change is to prefer call over the others, and specifically above impl default (this issue). I think this should be uncontroversial, as python semantics are are clear guide here -- a default is a *default*, if the call provides an explicit value then it should be preferred.

The second is to prefer spec default over impl default. The reason I think this makes sense is based on the use case (as far as I can think about -- if there's a use case I'm not considering, please let me know):
1. The use case for spec default is for the host application to add a new argument, while still accommodating old *callers*.
2. The use case for impl default is for the plugin to support both old and new spec (with & without argument).
The use case (2) suggests that impl default only covers for an old spec, and thus spec should win over it.

---

In previous discussions (e.g. https://github.com/pytest-dev/pluggy/issues/15#issuecomment-260109260), there was a suggestion to avoid the precedence considerations entirely by restricting the allowed default values to single sentinel, e.g. None. But IMO with clear documented precedence rules, there is no real need for this restriction.

**Comment by RonnyPfannschmidt** [comment]
i agree on the proposed order - call/spec/impl

i think its still a good idea to recommend using consistent sentinels and/or chosen optional to ensure a consistent and non-confusing ecosystem - but in some cases constants are just fine

**Comment by RonnyPfannschmidt** [comment]
in the initial ideas i recommendd hard limit to None/a sentinel - but currently i think its fine to allow wahts valid for functions

but with a recommendation for none/sentinels to keep things easy to comprehend
